csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -3814,150 +3730,212 @@ module ControlFlow {
+       * 2b. The set of splits for the successor is *always* empty.
+       *
+       * Only case 2a may introduce new sets of splits, so only predicates from
+       * this case are used in the definition of `TSplits`. 
+       *
+       * The predicates in this module are named after the cases above.
        */
-      private predicate succSplits2(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Splits succSplits, Completion c) {
-        succSplits1(pred, predSplits, succ, succSplits, c, 1)
-        or
-        succSplits0(pred, predSplits, succ, c) and
-        succSplits = TSplitsNil() and
-        not exists(SplitKind sk | sk.appliesToAlways(succ))
-      }
+      private module SuccSplits {
+        private predicate succReachable(Reachability::SameSplitsBlock b, ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Completion c) {
+          pred = b.getAnElement() and
+          b.isReachable(predSplits) and
+          succ = succ(pred, c)
+        }
 
-      /**
-       * Holds if
-       * - the set of splits `predSplits` applies to `pred`;
-       * - `succ` is a successor of `pred` with completion `c`;
-       * - the set of splits `succSplits` applies to `succ`;
-       * - each split in `succSplits` is either newly entered into, or passed
-       *   over from one of the predecessor splits in `predSplits`;
-       * - each split in `predSplits` (except possibly those in `predSplitsRemaining`)
-       *   is passed over to one of the successor splits in `succSplits`, or left; and
-       * - `succSplits` contains a split for each newly entered split.
-       */
-      private predicate succSplits3(ControlFlowElement pred, Splits predSplits, Splits predSplitsRemaining, ControlFlowElement succ, Splits succSplits, Completion c) {
-        succSplits2(pred, predSplits, succ, succSplits, c) and
-        predSplitsRemaining = predSplits and
-        // Enter a new split when required
-        forall(SplitInternal split |
-          split.hasEntry(pred, succ, c) |
-          split = succSplits.getASplit()
-        )
-        or
-        exists(SplitInternal predSplit |
-          succSplits3(pred, predSplits, TSplitsCons(predSplit, predSplitsRemaining), succ, succSplits, c) |
-          // Each predecessor split must be either passed over as a successor split,
-          // or must be left (possibly entering a new split)
-          predSplit.hasSuccessor(pred, succ, c) and
-          predSplit = succSplits.getASplit()
+        private predicate succSplits1b0(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Completion c) {
+          exists(Reachability::SameSplitsBlock b |
+            succReachable(b, pred, predSplits, succ, c) |
+            (succ = b.getAnElement() implies succ = b) and
+            not exists(SplitInternal split | split.hasEntry(pred, succ, c))
+          )
+        }
+
+        // A `forall` over the splits in `predSplits`, but written using explicit recursion
```

Comment:
Would be good to document the `forall` invariant - which I presume is that all the splits are the same?